File: src/lexer.cc
#include <nash/lexer.h>
#include <stdio.h>
#include <iostream>
#include <map>

const static std::map<std::string, TokenType> keywords = {
	{"and", TokenType::AND},
	{"else", TokenType::CLASS},
	{"false", TokenType::FALSE},
	{"for", TokenType::FOR},
	{"function", TokenType::FUN},
	{"if", TokenType::IF},
	{"or", TokenType::OR},
	{"return", TokenType::RETURN},
	{"true", TokenType::TRUE},
	{"var", TokenType::VAR},
	{"while", TokenType::WHILE},
	{"int", TokenType::INT_TYPE},
};


void Lexer::error(int line, int at, std::string msg){
	printf("Lexer::error::%d::%d: %s\n",line,at,msg.c_str());
}

std::vector<Token*> Lexer::scan(){
	while(!isAtEnd()){
		start = current;
		scanToken();
	}

	tokens.push_back(new Token(TokenType::_EOF, ""));
	
	return tokens;
}

bool Lexer::isAtEnd(){
	return (current >= (int)src.length());
}

void Lexer::scanToken(){
	char c = advance();
	switch(c){
		case ':':
			addToken(TokenType::COLON);
			break;
		case '(':
			addToken(TokenType::LEFT_PAREN);
			break;
		case ')':
			addToken(TokenType::RIGHT_PAREN);
			break;
		case '{': 
			addToken(TokenType::LEFT_BRACE); 
			break;
      	case '}': 
      		addToken(TokenType::RIGHT_BRACE); 
      		break;
      	case ',': 
      		addToken(TokenType::COMMA); 
      		break;
      	case '.': 
      		addToken(TokenType::DOT); 
      		break;
      	case '-': 
      		addToken(TokenType::MINUS); 
      		break;
      	case '+': 
      		addToken(TokenType::PLUS); 
      		break;
      	case ';': 
      		addToken(TokenType::SEMICOLON); 
      		break;
      	case '*': 
      		addToken(TokenType::STAR); 
      		break;
      	case '!':
	    	addToken(match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
	    	break;
	    case '=':
	    	addToken(match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
	    	break;
	    case '<':
	    	addToken(match('=') ? TokenType::LESS_EQUAL : TokenType::LESS);
	    	break;
	    case '>':
	    	addToken(match('=') ? TokenType::GREATER_EQUAL : TokenType::GREATER);
	    	break;
	    case '/':
	    	if (match('/')){
	    		while (peek() != '\n' && !isAtEnd()) advance();
	    	}
	    	else{
	    		addToken(TokenType::SLASH);
	    	}
	    	break;
	    case ' ':
      	case '\r':
      	case '\t':
        	break;
      	case '\n':
        	line++;
        	break;
        case '"':
        	string();
        	break;
      	default: 
      		if (isDigit(c)){
      			number();
      		}
      		else if (isAlpha(c)){
      			identifier();
      		}
      		else{
      			std::string error_msg = "Unexpected Character '";
      			error_msg.push_back(c);
      			error_msg.push_back('\'');
      			error(line, current, error_msg);
      		}
      		break; 
	}
}

char Lexer::advance(){
	return src[current++];
}

void Lexer::addToken(TokenType tt){
	addToken(tt, "");
}
void Lexer::addToken(TokenType tt, std::string token){
	tokens.push_back(new Token(tt, token));
	std::cout << "Token(" << (int)tt << "," << token << ")\n";
}

bool Lexer::match(char expected){
	if (isAtEnd()) return false;
    if (src[current] != expected) return false;

    current++;
    return true;
}

char Lexer::peek(){
	if (isAtEnd()) return '\0';
	return src[current];
}

void Lexer::string(){
	while (peek() != '"' && !isAtEnd()){
		if (peek() == '\n') line++;
		advance();
	}

	if (isAtEnd()){
		error(line, current, "Unterminated string.");
		return;
	}

	advance();

	std::string value = src.substr(start + 1, ((current-1) - (start+1)));
	addToken(TokenType::STRING, value);
}

bool Lexer::isDigit(char c){
	return c >= '0' && c <= '9';
}

void Lexer::number(){
	while (isDigit(peek())) advance();
	addToken(TokenType::NUMBER, src.substr(start, current-start));
}

char Lexer::peekNext(){
	if (current + 1 >= (int)src.length()) return '\0';
    return src[current + 1];
}

void Lexer::identifier(){
	 while (isAlphaNumeric(peek())) advance();
	 std::string text = src.substr(start, current-start);
	 auto _type = keywords.find(text);

	 if (_type == keywords.end()){
	 	addToken(TokenType::IDENTIFIER, text);
	 }
	 else{
	 	addToken(_type->second);
	 }
}

bool Lexer::isAlpha(char c){
	return (c >= 'a' && c <= 'z') ||
           (c >= 'A' && c <= 'Z') ||
            c == '_';
}

bool Lexer::isAlphaNumeric(char c){
	return isAlpha(c) || isDigit(c);
}

File: src/lexer.ts
import { LexerError, ThrowFatalError } from "./errors";

export enum TokenType {
    LPARAM = 1,
    RPARAM,
    STRING,
    INTEGER,
    FUNCTION,
    VARIABLE,
    CONST,
    IDENTIFIER,
    EQUAL,
    EQUALEQUAL,
    GREATER,
    LESS,
    LESS_OR_EQUAL,
    GREATER_OR_EQUAL,
    SEMICOLON,
    RETURN,
    PLUS,
    MINUS,
    STAR,
    SLASH,
    RSLASH,
    BANG,
    TRUE,
    FALSE,
    NULL,
    IF,
    ELSE,
    WHILE,
    FOR,
    CLASS,
    COLON,
    LCURLYBRACES,
    RCURLYBRACES,
    AND,
    OR,
    COMMA,
    EXTERN,
    IMPORT,
    AT,
    DEFER,
    STRUCT,
}
export interface Token {
    type: TokenType;
    value: any;
    typeName: string;
    line: number;
    column: number;
    isFunctionType: boolean;
    functionReturnType: string;
}

export class Lexer {
    input: string;
    index: number;
    currentCharacter: string = "";
    tokens: Array<Token> = [];
    currentLine: number = 1;
    currentColumn: number = 0;
    constructor(input: string) {
        this.input = input;
        this.index = 0;
    }
    addToken(
        token: TokenType,
        value: any = null,
        isFunctionType = false,
        functionReturnType = ""
    ) {
        this.currentColumn += value.length;
        this.tokens.push({
            type: token,
            value: value,
            typeName: TokenType[token],
            line: this.currentLine,
            column: this.currentColumn,
            isFunctionType: isFunctionType,
            functionReturnType: functionReturnType,
        });
    }
    updateCurrentCharacter() {
        this.currentCharacter = this.input.charAt(this.index);
    }
    getNextCharacter() {
        return this.input.charAt(this.index + 1);
    }
    getLastCharacter() {
        return this.input.charAt(this.index - 1);
    }
    goForward() {
        this.index++;
        this.updateCurrentCharacter();
    }
    atEnd() {
        return this.index === this.input.length - 1;
    }
    lex() {
        for (this.index = 0; this.index < this.input.length; this.index++) {
            this.updateCurrentCharacter();
            if (this.currentCharacter === "(") {
                this.addToken(TokenType.LPARAM, "(");
                continue;
            } else if (this.currentCharacter === ")") {
                this.addToken(TokenType.RPARAM, ")");
                continue;
            } else if (this.currentCharacter === ";") {
                this.addToken(TokenType.SEMICOLON, ";");
                continue;
            } else if (this.currentCharacter === ",") {
                this.addToken(TokenType.COMMA, ",");
                continue;
            } else if (this.currentCharacter === "+") {
                this.addToken(TokenType.PLUS, "+");
                continue;
            } else if (this.currentCharacter === "/") {
                if (this.getNextCharacter() === "/") {
                    this.goForward();
                    this.updateCurrentCharacter();
                    // @ts-ignore
                    while (this.currentCharacter !== "\n" && !this.atEnd()) {
                        this.goForward();
                        this.updateCurrentCharacter();
                    }
                    this.currentLine += 1;
                    this.currentColumn = 1;
                    continue;
                } else if (this.getNextCharacter() === "*") {
                    this.goForward();
                    this.goForward();
                    this.updateCurrentCharacter();
                    // @ts-ignore
                    while (this.currentCharacter !== "*" && !this.atEnd()) {
                        this.goForward();
                        this.updateCurrentCharacter();
                        // @ts-ignore
                        if (this.currentCharacter === "\n") {
                            this.currentLine += 1;
                            this.currentColumn = 1;
                        }
                    }
                    this.goForward();
                    this.updateCurrentCharacter();
                    if (this.currentCharacter === "/") {
                        continue;
                    } else {
                        throw new LexerError(
                            "Expected a '/' to finish the multiline comment"
                        );
                    }
                } else {
                    this.addToken(TokenType.RSLASH, "/");
                    continue;
                }
            } else if (this.currentCharacter === "-") {
                this.addToken(TokenType.MINUS, "-");
                continue;
            } else if (this.currentCharacter === "*") {
                this.addToken(TokenType.STAR, "*");
                continue;
            } else if (this.currentCharacter === "/") {
                this.addToken(TokenType.SLASH, "/");
                continue;
            } else if (this.currentCharacter === "!") {
                this.addToken(TokenType.BANG, "!");
                continue;
            } else if (this.currentCharacter === ":") {
                this.addToken(TokenType.COLON, ":");
                continue;
            } else if (this.currentCharacter === "{") {
                this.addToken(TokenType.LCURLYBRACES, "{");
                continue;
            } else if (this.currentCharacter === "@") {
                if (this.isLetter(this.getNextCharacter())) {
                    let identifier = this.currentCharacter;
                    while (true) {
                        if (
                            this.isLetter(this.getNextCharacter()) ||
                            // @ts-ignore
                            this.currentCharacter == "_" ||
                            this.isInteger(this.getNextCharacter())
                        ) {
                            this.goForward();
                            identifier += this.currentCharacter;
                        } else {
                            break;
                        }
                    }
                    this.addToken(TokenType.IDENTIFIER, identifier);
                } else {
                    this.addToken(TokenType.AT, "@");
                }
            } else if (this.currentCharacter === "}") {
                this.addToken(TokenType.RCURLYBRACES, "}");
                continue;
            } else if (this.currentCharacter === '"') {
                this.goForward();
                let str = "";
                while (true) {
                    if (this.currentCharacter != '"') {
                        str += this.currentCharacter;
                    } else {
                        if (this.getLastCharacter() === "\\") {
                            str = str.slice(0, -1);
                            str += '"';
                            this.goForward();
                            continue;
                        } else {
                            this.addToken(TokenType.STRING, str);
                            break;
                        }
                    }
                    if (this.atEnd()) {
                        throw new LexerError("Expected '\"'");
                    }
                    this.goForward();
                }
            } else if (this.currentCharacter === "=") {
                if (this.getNextCharacter() === "=") {
                    this.goForward();
                    this.addToken(TokenType.EQUALEQUAL, "==");
                } else {
                    this.addToken(TokenType.EQUAL, "=");
                }
                continue;
            } else if (this.currentCharacter === ">") {
                if (this.getNextCharacter() === "=") {
                    this.goForward();
                    this.addToken(TokenType.GREATER_OR_EQUAL, ">=");
                } else {
                    this.addToken(TokenType.GREATER, ">");
                }
                continue;
            } else if (this.currentCharacter === "<") {
                if (this.getNextCharacter() === "=") {
                    this.goForward();
                    this.addToken(TokenType.LESS_OR_EQUAL, "<=");
                } else {
                    this.addToken(TokenType.LESS, "<");
                }
                continue;
            } else if (this.isLetter(this.currentCharacter)) {
                let identifier = this.currentCharacter;
                while (true) {
                    if (
                        this.isLetter(this.getNextCharacter()) ||
                        this.currentCharacter == "_" ||
                        this.isInteger(this.getNextCharacter())
                    ) {
                        this.goForward();
                        identifier += this.currentCharacter;
                    } else {
                        break;
                    }
                }
                if (identifier === "function") {
                    if (this.getNextCharacter() === "<") {
                        this.goForward();
                        identifier += "<";
                        while (true) {
                            if (
                                (this.isLetter(this.getNextCharacter()) ||
                                    // @ts-ignore
                                    this.currentCharacter == "_" ||
                                    // @ts-ignore
                                    this.currentCharacter === "," ||
                                    this.isInteger(this.getNextCharacter())) &&
                                // @ts-ignore
                                this.currentCharacter != ">"
                            ) {
                                this.goForward();
                                identifier += this.currentCharacter;
                            } else {
                                break;
                            }
                        }
                        identifier += ">";
                        this.goForward();
                        this.addToken(TokenType.IDENTIFIER, identifier);
                    } else {
                        this.addToken(TokenType.FUNCTION, "function");
                    }
                } else if (identifier === "let") {
                    this.addToken(TokenType.VARIABLE, "let");
                } else if (identifier === "const") {
                    this.addToken(TokenType.CONST, "const");
                } else if (identifier === "return") {
                    this.addToken(TokenType.RETURN, "return");
                } else if (identifier === "null") {
                    this.addToken(TokenType.NULL, "null");
                } else if (identifier === "false") {
                    this.addToken(TokenType.FALSE, "false");
                } else if (identifier === "true") {
                    this.addToken(TokenType.TRUE, "true");
                } else if (identifier === "class") {
                    this.addToken(TokenType.CLASS, "class");
                } else if (identifier === "if") {
                    this.addToken(TokenType.IF, "if");
                } else if (identifier === "else") {
                    this.addToken(TokenType.ELSE, "else");
                } else if (identifier === "while") {
                    this.addToken(TokenType.WHILE, "while");
                } else if (identifier === "for") {
                    this.addToken(TokenType.FOR, "for");
                } else if (identifier === "and") {
                    this.addToken(TokenType.AND, "and");
                } else if (identifier === "or") {
                    this.addToken(TokenType.OR, "or");
                } else if (identifier === "external") {
                    this.addToken(TokenType.EXTERN, "external");
                } else if (identifier === "import") {
                    this.addToken(TokenType.IMPORT, "import");
                } else if (identifier === "defer") {
                    this.addToken(TokenType.DEFER, "defer");
                } else if (identifier === "struct") {
                    this.addToken(TokenType.STRUCT, "struct");
                } else {
                    this.addToken(TokenType.IDENTIFIER, identifier);
                }
                continue;
            } else if (this.isInteger(this.currentCharacter)) {
                let number = this.currentCharacter;
                while (true) {
                    if (this.isInteger(this.getNextCharacter())) {
                        this.goForward();
                        number += this.currentCharacter;
                    } else {
                        break;
                    }
                }
                this.addToken(TokenType.INTEGER, number);
                continue;
            } else if (this.currentCharacter === "\n") {
                this.currentLine += 1;
                this.currentColumn = 1;
            } else if (
                this.currentCharacter === " " ||
                this.currentCharacter === "\t" ||
                this.currentCharacter === "\r"
            ) {
                this.currentColumn += 1;
                continue;
            } else {
                ThrowFatalError(
                    "Unexpected character found",
                    this.currentLine,
                    this.currentColumn,
                    true
                );
            }
        }
    }
    isInteger(value: any) {
        return (
            !isNaN(value) &&
            (function (x) {
                return (x | 0) === x;
            })(parseFloat(value))
        );
    }
    isLetter(str: string) {
        return str.length === 1 && str.match(/[a-z]/i);
    }
}
